lib/Sema/CSSimplify.cpp

```diff
@@ -5075,6 +5075,34 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     }
   }
 
+  // In a l-value conversion involving an IUO base in a keypath application
+  // context we may endup matching a subscript member T? to T type. So in
+  // this case let's just strip optionals and attempt check if the types
+  // are bindable. Because a type U can be assigned to a subscript member of
+  // type U?
+  // e.g. a String value to a subscript member String?
+  //   func f(_ u: U!, _ kp: ReferenceWritableKeyPath<U, String?>, _ s: String)
+  //   {
+  //      u[keyPath: kp] = str
+  //   }
+  if (kind == ConstraintKind::Bind) {
+    auto last = locator.last();
+    if (last && last->is<LocatorPathElt::LValueConversion>()) {
+      SmallVector<LocatorPathElt, 4> path;
+      locator.getLocatorParts(path);
+      path.pop_back();
+
+      if (!path.empty() && path.back().is<LocatorPathElt::SubscriptMember>()) {
+        auto *loc = getConstraintLocator(locator, path);
+        if (desugar1->getOptionalObjectType() &&
+            !desugar2->getOptionalObjectType()) {
+          return matchTypes(desugar1->lookThroughAllOptionalTypes(), desugar2,
+                            kind, subflags, loc);
+        }
+      }
+    }
+  }
```

Comment:
I'm not sure this is the right fix – it seems like the problem is that we're attempting to bind the type variable for the result of the subscript too eagerly without first waiting to bind the type variable for the key path.

I think we'll want to add some logic in `getPotentialBindings` for a `KeyPathApplication` constraint similar to the logic we have for `ApplicableFunction` – where if we notice that the type variable is part of the callsite (on the left hand side), we mark the binding as being `FullyBound`, meaning that we'll try to delay binding the type variable until we've bound other type variables. This allows us to e.g try resolving the overload before attempting to bind the result type of a function.

In this case though, it looks like we only need to check whether the type variable appears in the `getThirdType()` of the `KeyPathApplication` (the projected result type), and if so mark the binding as `FullyBound`. This will allow us to wait until we've tried the key path type before attempting to bind the result type.

Reply:
Thanks for the review @hamishknight !

> I'm not sure this is the right fix – it seems like the problem is that we're attempting to bind the type variable for the result of the subscript too eagerly without first waiting to bind the type variable for the key path.

Yeah, I was really unsure about this approach ... but couldn't really think of anything else 

> the problem is that we're attempting to bind the type variable for the result of the subscript too eagerly without first waiting to bind the type variable for the key path.

That makes sense, and that is probably why we get a fixed type early on this case ... what doesn't happen for non-IUO optionals


> I think we'll want to add some logic in getPotentialBindings for a KeyPathApplication constraint similar to the logic we have for ApplicableFunction – where if we notice that the type variable is part of the callsite (on the left hand side), we mark the binding as being FullyBound, meaning that we'll try to delay binding the type variable until we've bound other type variables. This allows us to e.g try resolving the overload before attempting to bind the result type of a function.
In this case though, it looks like we only need to check whether the type variable appears in the getThirdType() of the KeyPathApplication (the projected result type), and if so mark the binding as FullyBound. This will allow us to wait until we've tried the key path type before attempting to bind the result type.

Great, I'll take a look on that, thanks again for the detailed comment :)

Reply:
Fixed @hamishknight, seems like I needed to better understand how the bindings work in those situations. 
Thanks again for the explanation :)

Reply:
Thanks @LucianoPAlmeida! I have one more comment but this LGTM!